Migrate AI gateway (#10420)
mlflow/mlflow: docs/source/llms/deployments/index.rst

.. _deployments:

========================================
MLflow Deployments Server (Experimental)
========================================

> The MLflow AI Gateway service is a powerful tool designed to streamline the usage and management of

Reviewer: are we going to replace these "AI Gateway" before merging similar to https://github.com/mlflow/mlflow/pull/10551 ?

Author: Yes, it's replaced now.

The gateway is designed to be flexible and adaptable, capable of easily defining and managing routes by updating the
configuration file. This enables the easy incorporation
of new LLM providers or provider LLM types into the system without necessitating changes to
applications that interface with the gateway. This level of adaptability makes the MLflow AI Gateway
Service an invaluable tool in environments that require agility and quick response to changes.

This simplification and centralization of language model interactions, coupled with the added
layer of security for API key management, make the MLflow AI Gateway service an ideal choice for
organizations that use LLMs on a regular basis.

.. toctree::
    :maxdepth: 1
    :hidden:

    guides/index

Tutorials and Guides
====================

If you're interested in diving right in to a step by step guide that will get you up and running with the MLflow AI Gateway 
as fast as possible, the guides below will be your best first stop. 

.. raw:: html

    <a href="guides/index.html" class="download-btn">View the AI Gateway Getting Started Guide</a><br/>

.. _deployments-quickstart:

> Quickstart

Reviewer: nit: are we going to remove this? (duplicated to the new tutorials)

Becoming familiar with these terms will help in configuring new endpoints (routes) and ease the use of the interface APIs for the AI Gateway.

.. _deployments-providers:

Providers
---------
The MLflow AI Gateway is designed to support a variety of model providers.
A provider represents the source of the machine learning models, such as OpenAI, Anthropic, and so on.
Each provider has its specific characteristics and configurations that are encapsulated within the model part of a route in the MLflow AI Gateway.

Supported Provider Models
~~~~~~~~~~~~~~~~~~~~~~~~~
The table below presents a non-exhaustive list of models and a corresponding route type within the MLflow AI Gateway.
With the rapid development of LLMs, there is no guarantee that this list will be up to date at all times. However, the associations listed
below can be used as a helpful guide when configuring a given route for any newly released model types as they become available with a given provider.
``N/A`` means that the provider or the AI Gateway implementation currently doesn't support the route type.

+--------------------------+--------------------------+--------------------------+--------------------------+

> |  Provider                | Routes                                                                         |

Reviewer: ```suggestion
|  Provider                | Route Types                                                                    |
```

Author: Fixed

| AI21 Labs                | - j2-ultra               | N/A                      | N/A                      |
|                          | - j2-mid                 |                          |                          |
|                          | - j2-light               |                          |                          |
+--------------------------+--------------------------+--------------------------+--------------------------+
| AWS Bedrock              | - Amazon Titan           | N/A                      | N/A                      |
|                          | - Third-party providers  |                          |                          |
+--------------------------+--------------------------+--------------------------+--------------------------+

† Llama 2 is licensed under the `LLAMA 2 Community License <https://ai.meta.com/llama/license/>`_, Copyright © Meta Platforms, Inc. All Rights Reserved.

Within each model block in the configuration file, the provider field is used to specify the name
of the provider for that model. This is a string value that needs to correspond to a provider the MLflow AI Gateway supports.

.. note::

>     `*` MLflow Model Serving will only work for chat or completions if the output return is in a route-compatible format. The

Reviewer: ```suggestion
    `*` MLflow Deployment Server will only work for chat or completions if the output return is in a route-compatible format. The
```

.. _deployments-providers:

Providers
---------
The MLflow Deployments Server is designed to support a variety of model providers.
A provider represents the source of the machine learning models, such as OpenAI, Anthropic, and so on.
Each provider has its specific characteristics and configurations that are encapsulated within the model part of an endpoint in the MLflow Deployments Server.

Supported Provider Models
~~~~~~~~~~~~~~~~~~~~~~~~~
The table below presents a non-exhaustive list of models and a corresponding endpoint type within the MLflow Deployments Server.
With the rapid development of LLMs, there is no guarantee that this list will be up to date at all times. However, the associations listed
below can be used as a helpful guide when configuring a given endpoint for any newly released model types as they become available with a given provider.
``N/A`` means that either the provider or the MLflow Deployments Server implementation currently doesn't support the endpoint type.

+--------------------------+--------------------------+--------------------------+--------------------------+

> |  Provider                | Endpoints                                                                      |

Reviewer: ```suggestion
|  Provider                | Endpoints  Types                                                               |
```

may not. If the model is not supported by the provider, the MLflow Deployments Server will return an HTTP 4xx error
when trying to route requests to that model.

.. important::

    Always check the latest documentation of the specified provider to ensure that the model you want
    to use is supported for the type of endpoint you're configuring.

Remember, the model you choose directly affects the results of the responses you'll get from the
API calls. Therefore, choose a model that fits your use-case requirements. For instance,
for generating conversational responses, you would typically choose a chat model.
Conversely, for generating embeddings of text, you would choose an embedding model.

.. _deployments_configuration:

Configuring the Deployments Server
==================================

The MLflow Deployments Server relies on a user-provided configuration file, written in YAML,
that defines the endpoints and providers available to the server. The configuration file dictates
how the deployments server interacts with various language model providers and determines the end-points that
users can access.

> Deployments Server Configuration

Reviewer: ```suggestion
Configuration YAML File
```
Original title is no different than the parent section title (L398).

update the endpoints with zero disruption or down time. This allows you to try out new providers or model types while keeping your applications steady and reliable.

In order to define an API key for a given provider, there are three primary options:

1. Directly include it in the YAML configuration file.
2. Use an environment variable to store the API key and reference it in the YAML configuration file.
3. Define your API key in a file and reference the location of that key-bearing file within the YAML configuration file.

If you choose to include the API key directly, replace ``$OPENAI_API_KEY`` in the YAML file with your
actual API key.

.. warning::

    The MLflow Deployments Server provides direct access to billed external LLM services. It is strongly recommended to restrict access to this server. See the section on :ref:`security <deployments_security>` for guidance.

If you prefer to use an environment variable (recommended), you can define it in your shell
environment. For example:

.. code-block:: bash

     export OPENAI_API_KEY="your_openai_api_key"

**Note:** Replace "your_openai_api_key" with your actual OpenAI API key.

.. _deployments_configuration_details:

> Deployments Server Configuration Details
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Detailed Configuration
----------------------
```

nit: To avoid too much nesting (less visibility in navigation bar).

+-------------------------------+----------------+----------+---------------+-------------------------------------------------------+
| **n**                         | integer        | No       | 1             | The number of chat completions to generate for the    |
|                               |                |          |               | specified prompt, between 1 and 5.                    |
+-------------------------------+----------------+----------+---------------+-------------------------------------------------------+
| **temperature**               | float          | No       | 0.0           | The sampling temperature to use, between 0 and 1.     |
|                               |                |          |               | Higher values will make the output more random, and   |
|                               |                |          |               | lower values will make the output more deterministic. |
+-------------------------------+----------------+----------+---------------+-------------------------------------------------------+

> | **max_tokens**                | integer        | No       | None          | The maximum completion length, between 1 and infinity |
> |                               |                |          |               | (unlimited).                                          |

Reviewer: ```suggestion
| **max_tokens**                | integer        | No       | None          | The maximum completion length, specify integer value >= 1. |
|                               |                |          |               | If None is specified, no maximum length limit is applied (unlimited).  |
```

nit: to make default (None) behavior clear

        )
        print(response)

LangChain Integration
~~~~~~~~~~~~~~~~~~~~~

`LangChain <https://github.com/langchain-ai/langchain>`_ supports `an integration for MLflow Deployments <https://python.langchain.com/docs/ecosystem/integrations/providers/mlflow>`_.
This integration enable users to use prompt engineering, retrieval augmented generation, and other techniques with LLMs in the deployments server.

.. code-block:: python
    :caption: Example

    import mlflow
    from langchain import LLMChain, PromptTemplate
    from langchain.llms import Mlflow

    llm = Mlflow(target_uri="http://127.0.0.1:5000", endpoint="completions")
    llm_chain = LLMChain(
        llm=llm,
        prompt=PromptTemplate(
            input_variables=["adjective"],
            template="Tell me a {adjective} joke",
        ),
    )
    result = llm_chain.run(adjective="funny")
    print(result)

    with mlflow.start_run():
        model_info = mlflow.langchain.log_model(llm_chain, "model")

    model = mlflow.pyfunc.load_model(model_info.model_uri)
    print(model.predict([{"adjective": "funny"}]))

.. _deployments_mlflow_models:

> MLflow Models

Reviewer: ```suggestion
Integrating with MLflow Models
```
